[webpack] setup lazy loading for all visualizations

williaster opened the pull request:
This PR adds lazy loading to all visualizations to partially address #4705 to reduce JS bundle size. This reduces the `explore` and `dashboard` bundle sizes by half, with the idea that the user would then fetch visualization bundles as needed, depending on what's present on a dashboard or as they toggle vis types in explore.

We want to do some empirical testing with some of our internal dashboards next week, and will update then. In the meantime feel free to pull the branch or comment on the code.

**before** 
<img src="https://user-images.githubusercontent.com/4496521/38128968-ca17523e-33b1-11e8-8595-c4cc099d295b.png" width="200" />

**after**
<img src="https://user-images.githubusercontent.com/4496521/38128995-f0df05e2-33b1-11e8-9225-18a0a73cc785.png" width="200" />

@graceguo-supercat @mistercrunch

williaster committed:
[webpack] setup lazy loading for all visualizations

williaster committed:
[lazy-load] push renderVis function to <Chart /> state

williaster committed:
no mapbox token

mistercrunch commented on superset/assets/visualizations/main.js:
```diff
@@ -51,53 +51,56 @@ export const VIZ_TYPES = {
+  [VIZ_TYPES.event_flow]: () => loadVis(import(/* webpackChunkName: "EventFlow" */ './EventFlow.jsx')),
+  [VIZ_TYPES.paired_ttest]: () => loadVis(import(/* webpackChunkName: "paired_ttest" */ './paired_ttest.jsx')),
+  [VIZ_TYPES.partition]: () => loadVis(import(/* webpackChunkName: "partition" */ './partition.js')),
+  [VIZ_TYPES.deck_scatter]: () => loadVis(import(/* webpackChunkName: "deckgl/layers/scatter" */ './deckgl/layers/scatter.jsx')),
```
can we use the same `webpackChunkName` for all `deck_*`?

williaster replied:
What's the goal here? for readability or to have all of that code in one chunk? 

I think using the same chunk name would only work if the same code was imported which is currently not possible since there are multiple index files for the various deck gl layers, right?

mistercrunch commented:
Wondering why the `webpackChunkName` doesn't show as such on entry file names, there's probably a way to hook that up.

We should probably have one `webpackChunkName` for all the visualizations that are mostly just `d3` and small libs.

mistercrunch commented:
But this is awesome btw!

williaster commented:
fyi hitting some issues testing this with prod data, still working on it. should be able to get the chunk files to have names that match the name in the code 👍

vibinsv09 commented:
@williaster Did the issue around testing with Prod data resolve?
Thanks for all the great work. This is awesome.

williaster commented:
sorry progress has been slower than expected, have been slammed with other projects including a focused sprint this week. will make this a priority to test early next week.

williaster committed:
[lazy loading] use native webpack import func to fix chunk names, add babel-plugin-syntax-dynamic-import, fix rebase bug.

williaster committed:
fix geojson import, undefined t, and fix async css bug

williaster committed:
[lazy load] actually add babel-plugin-syntax-dynamic-import

mistercrunch commented:
We should use the same webchunk name for all `deckgl`-related charts, same for `nvd3` and same all the ones that only require `d3`. I'm thinking we could have 4-5 bundles with all the visualizations in them.

Is it just a matter of sharing/reusing webchunknames?

williaster commented:
It's not as trivial as using the same chunk name. The final name consists of chunk name and hash. 

I'd rather do that optimization in another PR because I think you'd have to create an index file that exports all of deckgl and async import that file. I'm not as familiar with the deckgl code organization so you all might be best suited to do that?

williaster commented:
I can give deck.gl a shot if you want tho, I think the d3 bundle is almost over optimizing. That'll be harder to untangle potential totally as we refactor vis libraries.

mistercrunch commented:
I'm worried about the way things look here without bundling visualizations together. The sum of all visualization is much more than the original `explore` bundle.

A simple dashboard with a `filter_box` , nvd3, `table` and `pivot_table` will load like 24mb which is significantly more than the old explore bundle.

I'm guessing the load time improvements have to do with the fact that some of the work can be started upfront while the bundles are downloading.

mistercrunch commented:
We need to fix the fact that `filter_box` is 12mb. I'm thinking for visualizations that use `React`, they shouldn't have to import it and should somehow pick it up from `global`, same goes for `jquery` and perhaps a set of other things.

We probably need some sort of `global_imports` package that all apps (explore, dashboard, profile, ...) would import and build, and that visualization would pick from without importing/bundling.

williaster commented:
> I'm guessing the load time improvements have to do with the fact that some of the work can be started upfront while the bundles are downloading

Yes, and this is great from a user perspective because it doesn't block page render like it currently does. I think this will have the effect of making users feel like the page is loading _much_ faster.

> I'm thinking for visualizations that use React, they shouldn't have to import it and should somehow pick it up from global, same goes for jquery and perhaps a set of other things.

React [is not that large](https://bundlephobia.com/result?p=react@16.3.2) (even smaller if we could actually upgrade to `16.0.0`) and I can't tell if you mean literal globals (`window.jquery`) but we should move away from that pattern (and `jquery` in general, it's just syntactic sugar for JS and ideally we'd be using fetch not ajax) rather than embrace it. 

It seems like the proper way to pull out "globally shared" code is using the webpack `CommonChunksPlugin`, if chunks share code it will pull them into a new chunk. I used this to create a separate `chunk` for all shared deck.gl code, and I also added it in a way that it _should_ pull out all shared code into a separate chunk if it is imported by >=2 modules ... BUT this didn't appear to have any effect so the vis chunks must not really share that much code 🤷‍♀️ This is the new final webpack output (note the deckgl layer size decreases, and the new larger `deckgl.fb9e56a3bf69cd1ce139.chunk.js`):

```
Version: webpack 3.11.0
Time: 46756ms
                                                 Asset       Size                              Chunks                    Chunk Names
      deckgl/layers/grid.a8beccbf9f5b13e97f98.chunk.js    8.74 kB                                  32  [emitted]         deckgl/layers/grid
                  f4769f9bdb7466be65088239c12046d1.eot    20.1 kB                                      [emitted]
                 fa2772327f55d8198301fdb8bcfc8158.woff    23.4 kB                                      [emitted]
                  e18bbf611f2a2e43afc071aa2f4e1512.ttf    45.4 kB                                      [emitted]
                  89889688147bd7575d6327160d64e760.svg     109 kB                                      [emitted]
                nvd3_vis.7b1466ac696cd8d30467.chunk.js     6.8 MB                                   0  [emitted]  [big]  nvd3_vis
                  sankey.0da04cf564e3607b0313.chunk.js     268 kB                                   1  [emitted]  [big]  sankey
            paired_ttest.d9654f93c94179c43de4.chunk.js     304 kB                                   2  [emitted]  [big]  paired_ttest
             pivot_table.752772a0c72696357184.chunk.js    2.08 MB                                   3  [emitted]  [big]  pivot_table
    parallel_coordinates.0d0923b8adf4d520192a.chunk.js     204 kB                                   4  [emitted]         parallel_coordinates
               EventFlow.04abc2ccfab65027f920.chunk.js    2.57 MB                                   5  [emitted]  [big]  EventFlow
              big_number.39d33798fefe5e02632e.chunk.js    41.8 kB                                   6  [emitted]         big_number
                  markup.0ee56d948bd6cc1366fb.chunk.js    14.7 kB                                   7  [emitted]         markup
              filter_box.f76d5a45eff1c943db92.chunk.js      12 MB                                   8  [emitted]  [big]  filter_box
              time_table.49446835d78c8f19c439.chunk.js    1.51 MB                                   9  [emitted]  [big]  time_table
                  mapbox.b927c5a82390d9078f99.chunk.js    4.43 MB                                  10  [emitted]  [big]  mapbox
               partition.408cf680da8bcce8ad58.chunk.js     159 kB                                  11  [emitted]         partition
            deckgl/multi.bced534d9159ab18934d.chunk.js     512 kB  12, 13, 14, 29, 30, 31, 32, 33, 34  [emitted]  [big]  deckgl/multi
   deckgl/layers/scatter.ce1c712e5113e386fc6a.chunk.js     429 kB                                  13  [emitted]  [big]  deckgl/layers/scatter
deckgl/layers/screengrid.3619502e1991df96e5d5.chunk.js     400 kB                                  14  [emitted]  [big]  deckgl/layers/screengrid
                 heatmap.a587cf5f653617b1305a.chunk.js     108 kB                                  15  [emitted]         heatmap
                   table.48aed4719820c3e79b32.chunk.js    2.09 MB                                  16  [emitted]  [big]  table
             cal_heatmap.ed95b9fdabe40cb02ef1.chunk.js     348 kB                                  17  [emitted]  [big]  cal_heatmap
               world_map.3264a79e4259d761f4d0.chunk.js     941 kB                                  18  [emitted]  [big]  world_map
              word_cloud.477219c719efaab54f7c.chunk.js      45 kB                                  19  [emitted]         word_cloud
                    rose.92eafeac459006bad4bb.chunk.js    1.71 MB                                  20  [emitted]  [big]  rose
               histogram.d7ce608c93c52607b696.chunk.js    1.67 MB                                  21  [emitted]  [big]  histogram
                 treemap.f657e9f0a66f67c589ad.chunk.js      30 kB                                  22  [emitted]         treemap
                sunburst.5f07e304baedb2a39c0b.chunk.js    32.8 kB                                  23  [emitted]         sunburst
                 horizon.e8fba3db7283b0ab2024.chunk.js    18.9 kB                                  24  [emitted]         horizon
          directed_force.fcb6d26cd1fb2ae8add4.chunk.js    13.2 kB                                  25  [emitted]         directed_force
             country_map.c8e1683afb91b70582a1.chunk.js    16.7 kB                                  26  [emitted]         country_map
                   chord.d71d2fe59ff59c02121e.chunk.js    9.73 kB                                  27  [emitted]         chord
                  iframe.69c65c5f3fa42de0c8cf.chunk.js     1.8 kB                                  28  [emitted]         iframe
   deckgl/layers/polygon.3966937d1db2f1cdb4a1.chunk.js    6.75 kB                                  29  [emitted]         deckgl/layers/polygon
      deckgl/layers/path.47d1db5dd97d1e3f1535.chunk.js    7.42 kB                                  30  [emitted]         deckgl/layers/path
       deckgl/layers/hex.083fdc0e9eca06b50949.chunk.js    8.73 kB                                  31  [emitted]         deckgl/layers/hex
                448c34a56d699c29117adc64c43affeb.woff2      18 kB                                      [emitted]
   deckgl/layers/geojson.5e20b6a30b660329c019.chunk.js    12.1 kB                                  33  [emitted]         deckgl/layers/geojson
       deckgl/layers/arc.25cc2674046fd18c85a1.chunk.js    7.39 kB                                  34  [emitted]         deckgl/layers/arc
                 explore.69fb457561dcb2f21746.entry.js    25.6 MB                              35, 42  [emitted]  [big]  explore
                  sqllab.36c8988d9f36129a9183.entry.js    19.8 MB                              36, 42  [emitted]  [big]  sqllab
               dashboard.0a0c8aac20bdcac368aa.entry.js    15.4 MB                              37, 42  [emitted]  [big]  dashboard
                 profile.fa297a173d6a9ae45a79.entry.js    8.71 MB                              38, 42  [emitted]  [big]  profile
                addSlice.88e25db9764403cb2f42.entry.js    9.32 MB                              39, 42  [emitted]  [big]  addSlice
                 welcome.faa2156ccf68061b0f8a.entry.js    8.64 MB                              40, 42  [emitted]  [big]  welcome
                  deckgl.fb9e56a3bf69cd1ce139.chunk.js     8.7 MB                                  41  [emitted]  [big]  deckgl
                  common.32169c889bb0080def09.entry.js    3.99 MB                                  42  [emitted]  [big]  common
                   theme.489716a5e871d9900102.entry.js    6.15 kB                                  43  [emitted]         theme
```

The sad part is that this new deck.gl chunk doesn't load properly 😢 will keep digging a bit but may have to abandon it, or go talk to one of our web perf folks to see if they can help point me in the right direction. Sounds like there are many improvements for these exact thing in `webpack v4`, so we could consider upgrading to that (it will require new versions of node, so we'd need to specify `engine: { node: '^8.9.4' }` in the package.json, among other upgrades to webpack plugins, etc.)

williaster committed:
[webpack] working dev version of webpack v4

williaster commented:
all right things are going to get crazier from here, I was advised to not waste time getting the `CommonsChunksPlugin` to work and instead move to the webpack 4 hotness 🔥. I have a working (locally) version of that change, so could do that in another PR or could add it to this mix. @mistercrunch do you have any preference on that? they're both non-trivial changes I think, webpack v4 will require `"node": ">= 6.11.5 <7.0.0 || >= 8.9.0"` which means chef/backend config changes (cc @john-bodley)

Just for fun-sies I tried the changes here on top of the webpack 4 change and we are going going to generate a crazy number of bundles (because it does basically all of the shared chunk-ing you described above). [this is a good overview article](https://medium.com/webpack/webpack-4-code-splitting-chunk-graph-and-the-splitchunks-optimization-be739a861366).

NathanQuinn commented:
@williaster not sure if you are already familiar with this, but https://github.com/webpack-contrib/webpack-bundle-analyzer useful for deeper insights into the items impacting your bundle size here.  Beyond the deck.gl stuff, seems like brace, mathjs, and moment (locale) are most impactful.  There may be some opportunity to further optimize those, but would be interested in the webpack 4 upgrade stats before diving in on anything there.  Happy to help on this effort once you decide on direction to take this with @mistercrunch.

This is the output I got from your branch with the above plugin:

![superset_bundle_impact](https://user-images.githubusercontent.com/17075065/39667677-2621846e-5081-11e8-888b-78ee04001225.png)

vibinsv09 commented:
Hi,

Just wanted to check if we are planning to merge william's changes into superset?

Thanks

williaster commented:
Sorry I've been slow getting back to this, was on vacation and now trying to wrap up dashboard v2. I hit an issue with the chunk urls that are generated when I use the webpack v4 `splitChunks` logic. I still think this is going to be a huge win for performance, I'll try to spend some time on it again next week unless anyone else wants to pull this branch and PR into it? (let me know and I can push my webpack v4 changes)

williaster commented:
back on it, I think I figured out the webpack 4 issue. will post back today or tomorrow with bundle analysis etc! 🚀

williaster committed:
[webpack 4] fix url issues, use mini-css-extract-plugin and webpack-assets-manifest plugins

williaster committed:
[webpack 4] use splitchunks for all files, update templates to multi-file entrypoints

williaster commented:
Bundles are looking good in terms of maximum splitting/re-use. These images don't fully capture the picture for entry points (e.g., explorer, dashboard, sqllab) because every entry point "runtime" now actually consists of multiple scripts (for maximum sharing) [templates were updated accordingly]. 

Rough empirical #s for DOM loaded, MB transferred, etc. match pretty close to what I posted [above](https://github.com/apache/incubator-superset/pull/4727#issuecomment-385615006), but the number of requests is now higher. I'm not sure this really matters if empirical load time + total MB transferred are still good, though.

**before**
<img width="1436" alt="screen shot 2018-06-07 at 11 14 14 pm" src="https://user-images.githubusercontent.com/4496521/41142155-843db260-6aa9-11e8-8861-f06dae753e20.png">

<img width="348" alt="screen shot 2018-06-07 at 11 14 05 pm" src="https://user-images.githubusercontent.com/4496521/41142165-9112162a-6aa9-11e8-8308-a69f53f2c75a.png">


**after webpack 4 + splitchunks**
<img width="1436" alt="screen shot 2018-06-07 at 10 50 00 pm" src="https://user-images.githubusercontent.com/4496521/41142149-802906fc-6aa9-11e8-8958-c3575681fb43.png">

<img width="870" alt="screen shot 2018-06-07 at 11 14 50 pm" src="https://user-images.githubusercontent.com/4496521/41142179-97dc6028-6aa9-11e8-9335-9f44d84fe3d6.png">

mistercrunch commented:
Curious as to what happens if you pick multiple visualizations to have same `webpackChunkName` (trying to reduce the numder of chunks). Say all of the `deckgl` related ones or all of the ones that are just a few kbs already.

mistercrunch commented on superset/assets/package.json:
```diff
@@ -153,10 +156,10 @@
-    "webpack-manifest-plugin": "1.3.2",
-    "webworkify-webpack": "2.1.0"
+    "url-loader": "^1.0.1",
+    "webpack": "^4.6.0",
```
Oh noice! I tried to do this in the past and failed :)

williaster replied:
there were several loaders that had to be swapped out to support it 👍

mistercrunch commented:
Very excited to see this. Bundle sizes prior to this PR were starting to be quite embarrassing.

williaster committed:
[webpack 4] multiple theme entry files for markup vis css, don't uglify mapbox

williaster commented:
@mistercrunch re the same chunk for deckgl, I'm not sure that using the same chunk name is the best way to fix that given the options that the [SplitChunks](https://webpack.js.org/plugins/split-chunks-plugin/) plugin supports. For example you can specify a `minSize` criteria to be spun into a new chunk (default = `30kb`), `maxAsyncRequests`, or `maxInitialRequests`. 

My inclination is that we should see what happens with this since it's already a big change, and then tweak the optimizations from there.

Rebased on latest master just now and tried to test a lot myself locally. Will try to test on a staging box but if you could pull it and test more extensively with deck.gl that'd be super helpful! Unless we see any issues I think it's ready to merge!

williaster committed:
[webpack 4] lint python manifest changes, update yarn lock.

williaster committed:
[webpack 4] fix tests with babel-plugin-dynamic-import-node

mistercrunch commented:
Very nice. This seems to be a huge step forward, we can tweak things later as needed.

williaster commented:
These are the final set of 3x empirical tests per example dashboard before and after, on non-minimized code with hard refresh/no caching. They look **even better** than above!

- average total payload **decreased by 8.8 MB** (-65% change)
- DOM content load time **decreased 7.4s** (-66% change)
- Page total load time **decreased 8.5 s**
- the average # requests **increased by 13**

@mistercrunch shall we merge this? 🚀 

  | Total payload (MB) |   |  
-- | -- | -- | --
Dashboard | Before | After |  Delta
Births example | 12.5 | 3.9 | **-8.6 MB**
World data example | 12.5 | 3.9 | **-8.6 MB**
Misc charts example | 14.6 | 5.9 | **-8.7 MB**
Deck gl | 17.2 | 8.1 | **-9.1 MB**

  | DOM content loaded (seconds) |   |  
-- | -- | -- | --
Dashboard | Before | After | Delta
Births example | 11.52 | 3.84 | **-7.68 s**
World data example | 10.77 | 4.09 | **-6.68 s**
Misc charts example | 10.59 | 3.52 | **-7.07 s**
Deck gl | 11.69 | 3.62 | **-8.07 s**

  | Total time [all requests incl async] (sec) |   |  
-- | -- | -- | --
Dashboard | Before | After | Delta
Births example | 16.63 | 8.44 | **-8.19 s**
World data example | 13.17 | 5.51 | **-7.66 s**
Misc charts example | 17.58 | 9.5 | **-8.07 s**
Misc charts example w Deck gl | 18.43 | 8.31 | **-10.12 s**

  | Requests (count) |   |  
-- | -- | -- | --
Dashboard | Before | After | Delta
Births example | 32 | 44 | +12
World data example | 29 | 42 | +13
Misc charts example | 46 | 59 | +13
Deck gl (high bc of map tiles) | 196 | 208 | +12

williaster commented on superset/assets/src/chart/Chart.jsx:
```diff
@@ -81,6 +85,17 @@ class Chart extends React.PureComponent {
       );
     }
+
+    this.visPromise = visPromiseLookup[this.props.vizType];
```
I want to verify that this works properly when changing a vis in explorer, may need to move it out of `componentDidMount`

williaster replied:
confirmed this breaks, will fix it.

williaster committed:
[webpack 4] only use 'dynamic-import-node' plugin in tests, update <Chart /> vis promise when vis type changes

williaster committed:
[webpack 4] clean up package.json and yarn.lock after rebase

williaster commented:
I found and resolved an issue where the `dynamic-import-node` babel transpilation plugin broke webpack's bundle splitting capability ([see comment here](https://github.com/webpack/webpack/issues/5703#issuecomment-363267093)), which threw off my perf testing numbers above. They are now updated and look slightly better.

I also fixed the issue for updating visualizations promises and tested in explorer. @graceguo-supercat I also verified that switching between visualizations, or having multiple visualization of the same type does not result in new promises/requests.

Should be ready to merge if tests pass after the rebase I just did.

williaster committed:
[webpack 4] lint?

williaster committed:
[webpack 4] lint for real

williaster commented:
all right tests look good, should we merge @mistercrunch @graceguo-supercat @john-bodley ?

john-bodley commented:
:shipit:

williaster commented:
all righty 🤞🤞🤞
